File: stock-analyzer/scrapers/base_scraper.py
"""Base scraper class with common functionality."""

import requests
from bs4 import BeautifulSoup
import time
import random
from typing import Optional, Dict, Any
from abc import ABC, abstractmethod


class BaseScraper(ABC):
    """Base class for all financial data scrapers."""

    def __init__(self):
        self.session = requests.Session()
        self.session.headers.update({
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
            'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
            'Accept-Language': 'en-US,en;q=0.5',
            'Connection': 'keep-alive',
        })
        self.data = {}

    def fetch_page(self, url: str, retry_count: int = 3) -> Optional[BeautifulSoup]:
        """Fetch and parse a web page with retry logic."""
        for attempt in range(retry_count):
            try:
                # Add small random delay to avoid rate limiting
                time.sleep(random.uniform(0.5, 1.5))

                response = self.session.get(url, timeout=30)
                response.raise_for_status()

                return BeautifulSoup(response.content, 'lxml')

            except requests.RequestException as e:
                print(f"Attempt {attempt + 1} failed for {url}: {e}")
                if attempt < retry_count - 1:
                    time.sleep(2 ** attempt)  # Exponential backoff

        return None

    def clean_number(self, value: str) -> Optional[float]:
        """Clean and convert a string number to float."""
        if not value or value.strip() in ['-', '', 'N/A', 'NA']:
            return None

        try:
            # Remove commas, percentage signs, and whitespace
            cleaned = value.strip().replace(',', '').replace('%', '').replace('₹', '').replace('Cr', '').replace('cr', '')
            cleaned = cleaned.strip()

            # Handle negative numbers in parentheses
            if cleaned.startswith('(') and cleaned.endswith(')'):
                cleaned = '-' + cleaned[1:-1]

            return float(cleaned)
        except (ValueError, AttributeError):
            return None

    def parse_cr_value(self, value: str) -> Optional[float]:
        """Parse values in Crores (Cr) or Lakhs (L)."""
        if not value:
            return None

        value = value.strip().upper()
        multiplier = 1

        if 'CR' in value:
            value = value.replace('CR', '')
        elif 'L' in value or 'LAC' in value or 'LAKH' in value:
            value = value.replace('L', '').replace('LAC', '').replace('LAKH', '')
            multiplier = 0.01  # Convert lakhs to crores
        elif 'K' in value:
            value = value.replace('K', '')
            multiplier = 0.00001  # Convert thousands to crores

        num = self.clean_number(value)
        return num * multiplier if num is not None else None

    @abstractmethod
    def scrape(self, url: str) -> Dict[str, Any]:
        """Scrape data from the source. Must be implemented by subclasses."""
        pass

    @abstractmethod
    def get_company_name(self) -> str:
        """Get the company name from scraped data."""
        pass

File: stock-analyzer-java/src/main/java/com/stockanalyzer/scrapers/BaseScraper.java
package com.stockanalyzer.scrapers;

import com.stockanalyzer.models.FinancialData;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.Random;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Base class for all financial data scrapers.
 */
public abstract class BaseScraper {
    protected static final Logger logger = LoggerFactory.getLogger(BaseScraper.class);

    protected static final String USER_AGENT =
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36";

    protected static final int TIMEOUT_MS = 30000;
    protected static final int MAX_RETRIES = 3;

    protected FinancialData data;

    public BaseScraper() {
        this.data = new FinancialData();
    }

    /**
     * Fetch and parse a web page with retry logic.
     */
    protected Document fetchPage(String url) {
        for (int attempt = 1; attempt <= MAX_RETRIES; attempt++) {
            try {
                // Add small random delay to avoid rate limiting
                Thread.sleep(500 + new Random().nextInt(1000));

                Document doc = Jsoup.connect(url)
                        .userAgent(USER_AGENT)
                        .header("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8")
                        .header("Accept-Language", "en-US,en;q=0.5")
                        .timeout(TIMEOUT_MS)
                        .get();

                return doc;

            } catch (IOException e) {
                logger.warn("Attempt {} failed for {}: {}", attempt, url, e.getMessage());
                if (attempt < MAX_RETRIES) {
                    try {
                        Thread.sleep((long) Math.pow(2, attempt) * 1000);
                    } catch (InterruptedException ie) {
                        Thread.currentThread().interrupt();
                    }
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                logger.error("Thread interrupted while fetching {}", url);
            }
        }
        return null;
    }

    /**
     * Clean and convert a string number to Double.
     */
    protected Double cleanNumber(String value) {
        if (value == null || value.trim().isEmpty() ||
            value.equals("-") || value.equalsIgnoreCase("N/A") || value.equalsIgnoreCase("NA")) {
            return null;
        }

        try {
            // Remove commas, percentage signs, currency symbols, and whitespace
            String cleaned = value.trim()
                    .replace(",", "")
                    .replace("%", "")
                    .replace("₹", "")
                    .replace("Cr", "")
                    .replace("cr", "")
                    .replace("Rs.", "")
                    .replace("Rs", "")
                    .trim();

            // Handle negative numbers in parentheses
            if (cleaned.startsWith("(") && cleaned.endsWith(")")) {
                cleaned = "-" + cleaned.substring(1, cleaned.length() - 1);
            }

            return Double.parseDouble(cleaned);
        } catch (NumberFormatException e) {
            return null;
        }
    }

    /**
     * Parse values with Cr (Crores) or L (Lakhs) suffix.
     */
    protected Double parseCroreValue(String value) {
        if (value == null) return null;

        String upper = value.toUpperCase().trim();
        double multiplier = 1.0;

        if (upper.contains("CR")) {
            upper = upper.replace("CR", "");
        } else if (upper.contains("LAKH") || upper.contains("LAC") || upper.contains("L")) {
            upper = upper.replaceAll("LAKH|LAC|L", "");
            multiplier = 0.01; // Convert lakhs to crores
        } else if (upper.contains("K")) {
            upper = upper.replace("K", "");
            multiplier = 0.00001; // Convert thousands to crores
        }

        Double num = cleanNumber(upper);
        return num != null ? num * multiplier : null;
    }

    /**
     * Extract number from text using regex.
     */
    protected Double extractNumber(String text) {
        if (text == null) return null;

        Pattern pattern = Pattern.compile("-?[\\d,]+\\.?\\d*");
        Matcher matcher = pattern.matcher(text.replace(",", ""));

        if (matcher.find()) {
            try {
                return Double.parseDouble(matcher.group());
            } catch (NumberFormatException e) {
                return null;
            }
        }
        return null;
    }

    /**
     * Abstract method to scrape data from a URL.
     */
    public abstract FinancialData scrape(String url);

    /**
     * Get the scraped data.
     */
    public FinancialData getData() {
        return data;
    }
}
